com.unity.ml-agents/Runtime/Inference/BarracudaModelExtensions.cs

```diff
@@ -7,6 +9,140 @@ namespace Unity.MLAgents.Inference
+            {
+                names.Add(input.name);
+            }
+
+            foreach (var mem in model.memories)
+            {
+                names.Add(mem.input);
+            }
+
+            names.Sort();
+
+            return names.ToArray();
+        }
+
+        /// <summary>
+        /// Generates the Tensor inputs that are expected to be present in the Model.
+        /// </summary>
+        /// <param name="model">
+        /// The Barracuda engine model for loading static parameters.
+        /// </param>
+        /// <returns>TensorProxy IEnumerable with the expected Tensor inputs.</returns>
+        public static IReadOnlyList<TensorProxy> GetInputTensors(this Model model)
+        {
+            var tensors = new List<TensorProxy>();
+
+            if (model == null)
+                return tensors;
+
+            foreach (var input in model.inputs)
+            {
+                tensors.Add(new TensorProxy
+                {
+                    name = input.name,
+                    valueType = TensorProxy.TensorType.FloatingPoint,
+                    data = null,
+                    shape = input.shape.Select(i => (long)i).ToArray()
+                });
+            }
+
+            foreach (var mem in model.memories)
+            {
+                tensors.Add(new TensorProxy
+                {
+                    name = mem.input,
+                    valueType = TensorProxy.TensorType.FloatingPoint,
+                    data = null,
+                    shape = TensorUtils.TensorShapeFromBarracuda(mem.shape)
+                });
+            }
+
+            tensors.Sort((el1, el2) => el1.name.CompareTo(el2.name));
+
+            return tensors;
+        }
+
+
+        /// <summary>
+        /// Get number of visual observation inputs to the model.
+        /// </summary>
+        /// <param name="model">
+        /// The Barracuda engine model for loading static parameters.
+        /// </param>
+        /// <returns>Number of visual observation inputs to the model</returns>
+        public static int GetNumVisualInputs(this Model model)
+        {
+            var count = 0;
+            if (model == null)
+                return count;
+
+            foreach (var input in model.inputs)
+            {
+                if (input.name.StartsWith(TensorNames.VisualObservationPlaceholderPrefix))
+                {
+                    count++;
+                }
+            }
+
+            return count;
+        }
+
+        /// <summary>
+        /// Get array of the output tensor names of the model.
+        /// </summary>
+        /// <param name="model">
+        /// The Barracuda engine model for loading static parameters.
+        /// </param>
+        /// <returns>Array of the output tensor names of the model</returns>
+        public static string[] GetOutputNames(this Model model)
+        {
+            var names = new List<string>();
+
+            if (model == null)
```

Comment:
I'm not quite sure how nulls work with extension methods - may need to figure out if this was being hit anywhere, and add a null check there instead.

Reply:
From my experiments it works just the same as calling the method directly with a null reference passed in, and these are not triggered anywhere as far as I know.
No model given plus inference type == inference_only is the only case I found the model could be null, and it's already captured properly.